nixos: unify virtual console options

rnhmjoj opened the pull request:
###### Motivation for this change

As per commit message:

This commit moves all the virtual console related options
to a dedicated config/console.nix NixOS module.

Currently most of these are defined in config/i18n.nix
with a "console" prefix like `i18n.consoleFont`,
`i18n.consoleColors` or under `boot` and are implemented
in tasks/kbd.nix.
Since they have little to do with actual internationalisation
and are (informally) in an attrset already, it makes sense to
move them to a specific module.

Also, this commit changes the console colors implementation
to use the kernel parameters instead of relying on terminal
escape sequences. This means the palette is applied by the
kernel itself with no for custom code running in the initrd
and works for virtual terminals (not only tty0).


###### Things done

- [ ] Tested via one or more NixOS tests (not aware of any one specific)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @vcunat @abbradar @greedy @the-kenny

rnhmjoj committed:
nixos: unify virtual console options

This commit moves all the virtual console related options
to a dedicated config/console.nix NixOS module.

Currently most of these are defined in config/i18n.nix
with a "console" prefix like `i18n.consoleFont`,
`i18n.consoleColors` or under `boot` and are implemented
in tasks/kbd.nix.
Since they have little to do with actual internationalisation
and are (informally) in an attrset already, it makes sense to
move them to a specific module.

rnhmjoj committed:
docs/release-notes: document console related changes

rnhmjoj committed:
docs/release-notes: fix indentation

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review-may-2019/3032/80

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/what-are-your-goals-for-20-03/4773/4

pstch commented:
- [ ] changes are backward compatible : _renamed options_
- [X] renamed options are declared with `mkRenamedOptionModule`
- [X] changes that are not backward compatible are documented in release notes
- [ ] module tests succeed on linux amd64 _not applicable_
- [X] options types are appropriate : _no change_
- [X] options description is set : _no change_
- [X] options example is provided : _no change_
- [X] documentation affected by the changes is updated : _manual not affected_

Suggestions:
 - as this PR changes important options that are defined in many configurations, it may be better to isolate it from the changes to the color palette application

rnhmjoj commented:
Thank you for the review, @pstch.

> as this PR changes important options that are defined in many configurations, it may be better to isolate it from the changes to the color palette application

I thought about it but I made these changes simultaneously so it would require some nontrivial history rewriting.

pstch commented:
I missed something in my review:
 - The old definition of `i18n.consoleKeyMap` has not been removed (in `i18n.nix`)
 - The XML of the release note is broken.

I'm rebasing your commits, fixing both of these issues, and will push my branch so that you can use it if you want. I also removed the changes to the color palette application by diffing the two files, so that we can add them in another PR.

pstch commented:
@rnhmjoj I [rebased your PR](https://github.com/pstch/nixpkgs/tree/feature/unified-console-options) on recent master, applying the following steps:
 - fixing the XML in the release notes, removing the commit for indentation that doesn't apply anymore
 - removing the changes to the color palette application
 - removing the old definition of `i18n.consoleKeyMap` in `i18n.nix`

I have tested these commits on my machine by rebasing them on 19.09.

rnhmjoj committed:
nixos/console: set colors using kernel parameters

This commit changes the console colors implementation
to use the kernel parameters instead of relying on terminal
escape sequences. This means the palette is applied by the
kernel itself with no custom code running in the initrd
and works for all virtual terminals (not only tty0).

rnhmjoj commented:
@pstch Thank you. I fixed the merge conflict and followed your suggestions.

rnhmjoj commented:
Fixed yet another conflict.

rnhmjoj commented:
This time it was me. I wonder why git can't automaticallt merge this...

Infinisil commented on nixos/modules/config/console.nix:
```diff
@@ -0,0 +1,203 @@
+          ${if substring 0 1 cfg.font == "/" then ''
+            font="${cfg.font}"
+          '' else ''
+            font="$(echo ${consoleEnv}/share/consolefonts/${cfg.Font}.*)"
```
Should be `cfg.font`

rnhmjoj replied:
Thank you, I missed it.

rnhmjoj commented:
@Infinisil What to do you think about adding a new top level attribute? I'm not too keen but I didn't find a suitable existing one.

infinisil reviewed (approved):
Sounds like a good idea to me, it really doesn't fit anywhere else.

I tested this in a VM, seems to work fine, let's merge :)

rnhmjoj commented:
Thank you!
